### PR TITLE
types: optional field redaction support

### PR DESCRIPTION
**Pull request by tsenart**
This adds support for redacting optional fields in external service
configs, particularly Pagure's optional token field.

Part of #28028

### CODE REVIEW

File: internal/types/secret_test.go [hunk 1]
```diff
@@ -65,6 +65,9 @@ func TestRoundTripRedactExternalServiceConfig(t *testing.T) {
 			Dependencies: []string{"placeholder"},
 		},
 	}
+	pagureConfig := schema.PagureConnection{
+		Url: "https://src.fedoraproject.org",
+	}
```
Comment (varungandhi-src) [inline internal/types/secret_test.go:70]:
```suggestion
	pagureConfig := schema.PagureConnection{
		Url: "https://src.fedoraproject.org",
		Token: "top secret token",
	}
```

The current code will test the `""` string code path, whereas I think what you want to test here is that the token is redacted correctly if it is present.

File: internal/types/secret.go [hunk 2]
```diff
@@ -166,7 +169,8 @@ func unredactFields(old, new string, cfg interface{}) (string, error) {
 	for _, field := range jsonStringFields {
 		v, err := jsonc.ReadProperty(new, field.path...)
 		if err != nil {
-			return new, err
+			// This field was deleted, so we skip any edits to it.
+			continue
```
Comment (varungandhi-src) [inline internal/types/secret.go:173]: Could you clarify why this is needed? I don't think redacting shouldn't delete any fields, it only modifies the value. So if a field was deleted, it seems like something has gone wrong.

**Review by varungandhi-src — APPROVED** [review]

File: internal/types/secret.go (hunk 2 above)
Comment (tsenart) [reply internal/types/secret.go:173]: This is covering deletion of a field in the new blob, that was previously redacted.

**Comment by varungandhi-src** [comment]
I misunderstood the PR; discussed over Slack. Thanks for adding the comment. LGTM.

**Pull request merged**